templates/_helpers.tpl

```diff
@@ -162,17 +166,27 @@ postgres://{{ template "harbor.database.username" . }}:{{ template "harbor.datab
   {{- end }}
 {{- end -}}
 
+{{- define "harbor.redis.username" -}}
+  {{- with .Values.redis }}
+    {{- if ne .type "internal" -}}
+      {{- ternary ((lookup "v1" "Secret" $.Release.Namespace .external.existingSecret).data.REDIS_USERNAME | b64dec) .external.username (not (not .external.existingSecret)) }}
+    {{- end -}}
+  {{- end }}
+{{- end -}}
+
 {{- define "harbor.redis.password" -}}
   {{- with .Values.redis }}
-    {{- ternary "" .external.password (eq .type "internal") }}
+    {{- if ne .type "internal" -}}
+      {{- ternary ((lookup "v1" "Secret" $.Release.Namespace .external.existingSecret).data.REDIS_PASSWORD | b64dec) .external.password (not (not .external.existingSecret)) }}
+    {{- end -}}
   {{- end }}
 {{- end -}}
 
-/*scheme://[:password@]host:port[/master_set]*/
+/*scheme://[username][:password@]host:port[/master_set]*/
 {{- define "harbor.redis.url" -}}
   {{- with .Values.redis }}
     {{- $path := ternary "" (printf "/%s" (include "harbor.redis.masterSet" $)) (not (include "harbor.redis.masterSet" $)) }}
-    {{- $cred := ternary (printf "%s:%s@" (.external.username | urlquery) (.external.password | urlquery)) "" (and (eq .type "external" ) (not (not .external.password))) }}
+    {{- $cred := ternary (printf "%s:%s@" ((include "harbor.redis.username" $) | urlquery)) ((include "harbor.redis.password" $) | urlquery)) "" (and (eq .type "external" ) (not (not (include "harbor.redis.password" $)))) }}
```

Comment:
helm lint is not happy with this change. You added one `)`to much after the first `urlquery`.

Should look like this:
```
    {{- $cred := ternary (printf "%s:%s@" ((include "harbor.redis.username" $) | urlquery) ((include "harbor.redis.password" $) | urlquery)) "" (and (eq .type "external" ) (not (not (include "harbor.redis.password" $)))) }}
```